File: widget_canvas/widget_canvas.py

from __future__ import division, print_function, unicode_literals, absolute_import

import os
import base64
import uuid

import numpy as np

import IPython
from ipywidgets import widgets
import traitlets

from . import image


def inject():
    """
    Inject JavaScript component into browser front end.
    """
    fname = 'widget_canvas.js'
    path_module = os.path.abspath(os.path.dirname(__file__))

    f = os.path.join(path_module, fname)
    js = IPython.display.Javascript(filename=f)  # data=None, url=None, filename=None, lib=None

    IPython.display.display(js)

# Inject JavaScript when module is first loaded.
inject()


class CanvasImage(widgets.widget.DOMWidget):
    """
    Display images using HTML5 Canvas with IPython Notebook widget system.
    """
    # _model_name:   Name of Backbone model registered in the front-end to create and sync with.
    # _model_module: A requirejs module name in which to find _model_name.
    # _view_name:    Default view registered in the front-end to use to represent the widget.
    # _view_module:  A requirejs module in which to find _view_name.

    _view_name = traitlets.Unicode('CanvasImageView', sync=True)

    # Encoded image data and metadata
    _encoded = traitlets.Bytes(help='Encoded image data', sync=True)
    _format = traitlets.Unicode(help='Image encoding format', sync=True)

    # HTML/DOM/CSS screen display width and height
    height = traitlets.CInt(help='image screen display height', sync=True)
    width = traitlets.CInt(help='image screen display width', sync=True)

    # Image rendering quality. https://developer.mozilla.org/en/docs/Web/CSS/image-rendering
    pixelated = traitlets.Bool(False, help='Render images via nearest neighbor sampling',
                               sync=True)

    # Canvas width and height, slaved to image data width and height.
    width_canvas = traitlets.CInt(help='image data width', sync=True)
    height_canvas = traitlets.CInt(help='image data height', sync=True)

    # Mouse event information.
    _mouse_active = traitlets.Bool(False, help='Indicate if mouse events are active', sync=True)
    _mouse_event = traitlets.Dict(help='Front-end mouse event information', sync=True)

    _uuid = traitlets.Unicode(uuid.uuid1().hex, help='widget unique ID string', sync=True)
    _something = traitlets.Bool(True, sync=True)

    def __init__(self, data=None, url=None, format='webp', quality=70, force_js=False,
                 **kwargs):
        """
        Instantiate a new Canvas Image Widget object.

        Display images using HTML5 Canvas with IPython Notebook widget system. Keyword `value`
        contains the input image data.  It must be a Numpy array (or equivalent) with a shape
        similar to one of the following:
            (rows, columns)    - Greyscale
            (rows, columns, 1) - Greyscale
            (rows, columns, 3) - RGB
            (rows, columns, 4) - RGBA

        If data type is neither of np.uint8 or np.int16, it will be cast to uint8 by scaling
        min(data) -> 0 and max(data) -> 255.

        If you supply a URL that points to an image then that image will be fetched and stored
        locally as a uint8 byte array.
        """
        super(CanvasImage, self).__init__(**kwargs)

        if force_js:
            inject()

        # Internal Python handler for JS mouse events synced through _mouse_event traitlet
        # https://developer.mozilla.org/en-US/docs/Web/Reference/Events
        # https://developer.mozilla.org/en-US/docs/Web/API/MouseEvent
        self.on_trait_change(self._handle_mouse, str('_mouse_event'))  # JS --> Python

        # Allow user to attach Python callback functions to operate in response to mouse events.
        self._mouse_event_dispatchers = {}  # Python (this module) --> user's Python function(s)
        for kind in ['move', 'up', 'down', 'click', 'wheel']:
            # Independent dispatcher for each 'kind' of event.
            self._mouse_event_dispatchers[kind] = widgets.widget.CallbackDispatcher()

        # Store init data in traitlet(s)
        self.format = format
        self.quality = quality

        if url:
            # Fetch image from URL and decompress into local numpy array
            data_comp, format_orig = image.download(url)
            data = image.decompress(data_comp)

        # Set image data
        self.data = data

    def close(self):
        print('close!!')
        super().close()

    @property
    def data(self):
        """
        Image data stored as a numpy array.
        """
        return self._data

    @data.setter
    def data(self, data):
        with self.hold_sync():
            # Pause syncing of state changes until the context manager is released
            if issubclass(type(data), np.ndarray):
                HxW = data.shape[:2]

                # Compress input image data and encode via Base64
                self._data = image.setup_data(data)

                data_comp = image.encode(self._data, fmt=self.format)
                data_encode = base64.b64encode(data_comp)
            else:
                # Clobber image data
                self._data = None
                HxW = 0, 0
                # data_comp = None
                data_encode = b''

            # Update traitlets
            self._encoded = data_encode
            self.height, self.width = HxW
            self.height_canvas, self.width_canvas = HxW

    @property
    def format(self):
        """
        Image compression/encoding format, e.g. 'jpg', 'png', 'webp', etc.
        """
        return self._format

    @format.setter
    def format(self, value):
        valid_formats = ['png', 'jpg', 'jpeg', 'webp']
        if value.lower() not in valid_formats:
            raise ValueError('Invalid encoding format: {}'.format(value))

        self._format = value.lower()

    def display(self):
        """
        Display image to Notebook using IPython's rich-display infrastructure.
        """
        IPython.display.display(self)

    #####################################################
    # Respond to mouse events by calling registered Python event handlers
    def _scale_xy_screen_to_data(self, event):
        """
        Event XY values are generated in screen coordinates.  This function convert the values
        to data XY coordinates.
        """
        event['canvasY'] = round(event['canvasY'] * self._data.shape[0] / self.height)
        event['canvasX'] = round(event['canvasX'] * self._data.shape[1] / self.width)

        return event

    def _handle_mouse(self, name_trait, event):  # info_old, info_new):
        """
        Respond to JavaScript mouse events by calling registered Python function event handlers
        https://developer.mozilla.org/en-US/docs/Web/API/MouseEvent
        """
        # Event types in JavaScript: 'mousemove', 'mouseup', 'mousedown', 'click', 'wheel'
        # Event types in Python:          'move',      'up',      'down', 'click', 'wheel'
        # kind_js_to_py = {'mousemove': 'move',
        #                  'mouseup': 'up',
        #                  'mousedown': 'down',
        #                  'click': 'click',
        #                  'wheel': 'wheel'}

        # Preprocess event data
        # event['type'] = kind_js_to_py[event['type']]  # update event types
        event = self._scale_xy_screen_to_data(event)

        # Handle the event
        try:
            kind = event['type']
            self._mouse_event_dispatchers[kind](self, event)
        except KeyError:
            raise ValueError('Really unexpected event type: {}'.format(kind))

    #######################################################
    # Register Python fnctions as mouse event handlers
    #
    #   callback : function to be called with two arguments: widget instance and event information.
    #   remove : bool (optional), set to true to unregister the callback function.
    #
    #   https://developer.mozilla.org/en-US/docs/Web/API/MouseEvent
    #   https://developer.mozilla.org/en-US/docs/Web/Events/wheel
    def _num_handlers(self):
        """
        Number of registered mouse event handlers.
        """
        numbers = [len(d.callbacks) for d in self._mouse_event_dispatchers.values()]
        return sum(numbers)

    def on_mouse(self, callback, kinds=None, remove=False):
        """
        Register a callback function for one or more kinds of mouse events.
        Valid mouse event kinds: 'move', 'up', 'down', 'click', 'wheel'

        Default kinds=None will register the callback with all mouse event types.
        Setting kind to a string or a list of strings will register the callback with
        just the specified kinds.

        Set keyword remove=True in order to unregister an existing callback function.
        Arguments to the supplied callback function are the widget instance and event information.
        """
        # Default to all defined event dispatchers
        if not kinds:
            # Default all mouse event kinds: 'move', 'up', 'down', 'click', 'wheel'
            kinds = self._mouse_event_dispatchers.keys()

        if isinstance(kinds, str):
            kinds = [kinds]

        for k in kinds:
            if k not in self._mouse_event_dispatchers:
                raise ValueError('Invalid callback type: {}'.format(k))

            # Register/un-register user-defined callback function
            self._mouse_event_dispatchers[k].register_callback(callback, remove=remove)

        # Enable/disable mouse event handling at front end.
        self._mouse_active =  self._num_handlers() > 0

    def on_mouse_move(self, callback):
        self.on_mouse(callback, 'move')

    def on_mouse_up(self, callback):
        self.on_mouse(callback, 'up')

    def on_mouse_down(self, callback):
        self.on_mouse(callback, 'down')

    def on_mouse_click(self, callback,):
        self.on_mouse(callback, 'click')

    def on_mouse_wheel(self, callback):
        self.on_mouse(callback, 'wheel')

    def unregister(self):
        """
        Unregister all mouse event handler functions.
        """
        callbacks = []
        for dispatcher in self._mouse_event_dispatchers.values():
            callbacks += dispatcher.callbacks

        for cb in callbacks:
            self.on_mouse(cb, remove=True)

        # Enable/disable mouse event handling at front end.
        self._mouse_active =  self._num_handlers() > 0

#################################################

if __name__ == '__main__':
    pass

File: widget_canvas/widget_canvas.js
require(["widgets/js/widget", "widgets/js/manager"], function(widget, manager) {


    /////////////////////////////////////////////////
    // Helper functions

    function findAncestor(elem, cls) {
        // http://stackoverflow.com/questions/22119673/find-the-closest-ancestor-element-that-has-a-specific-class
        while ((elem = elem.parentElement) && !elem.classList.contains(cls));
        return elem;
    }

    /////////////////////////////////////////////////

    // define(function (require) {
    //     var widget = require('widgets/js/widget');

    var CanvasImageView = widget.DOMWidgetView.extend({
        render: function() {
            // Backbone Model --> JavaScript View
            // Render a widget's view instance to the DOM.

            // This project's view is quite simple: just a single <canvas> element.
            // if (this.model.get('canvas_id') == undefined) {
            //     this.model.set('canvas_id', guid());
            // }
            // console.log(this.model.get('_uuid'));
            this.setElement('<canvas />');

            this.canvas = this.el
            this.canvas.className = 'my_canvas ' + this.model.get('_uuid')

            // Dedicated event handler(s) for some special cases
            // http://backbonejs.org/#Events-on
            this.model.on('change:_encoded_url', this.update_encoded, this);
            this.model.on('change:_something', this.something, this);

            // Internal image object serving to render new image src data.  This object will
            // later be used as source data argument to the canvas' own `drawImage()` method.
            this.imageWork = new Image();
            var that = this
            this.imageWork.onload = function() {
                that.draw()
            }

            // Ideas for handling keyboard events in the future:
            // http://stackoverflow.com/questions/3729034/javascript-html5-capture-keycode-and-write-to-canvas

            // Mouse event throttle
            this._mouse_timestamp = 0
            this._mouse_time_threshold = 50 // milliseconds

            // Prevent mouse cursor from changing to text selection mode.
            // http://stackoverflow.com/a/11805438/282840
            // when-i-click-on-a-canvas-and-drag-my-mouse-the-cursor-changes-to-a-text-selecti
            this.canvas.onmousedown = function(event) {
                event.preventDefault();
            };

            // Prevent page from scrolling with mouse wheel events over canvas
            // http://stackoverflow.com/questions/10313142/
            //        javascript-capture-mouse-wheel-event-and-do-not-scroll-the-page
            this.canvas.onwheel = function(event) {
                event.preventDefault();
            };

            // Prevent context menu popup from right-click on canvas
            this.canvas.oncontextmenu = function(event) {
                event.preventDefault();
            };

            this.update();
            this.update_encoded();
        },

        update: function() {
            // Python --> JavaScript (generic)
            // Copy new value from Backbone model, apply to this View.
            // This method handles updates for almost everything except a select few traitlets that
            // have dedicated update functions.

            // Currently canvas width and height are slaved (on the Pythn side) to image's
            // inherent width and height.  Later this might be upgraded to support canvas' builtin
            // affine transform functions.

            // Awesome article about resizing canvas and/or displayed element
            // http://webglfundamentals.org/webgl/lessons/webgl-resizing-the-canvas.html

            // Update canvas width and height.
            // if (this.model.get('_width_canvas') !== undefined) {
            this.canvas.width = this.model.get('_width_canvas')
                // } else {
                // https://developer.mozilla.org/en-US/docs/Web/API/Element/removeAttribute
                // this.canvas.removeAttribute('_width_canvas');
                // }

            // if (this.model.get('_height_canvas') !== undefined) {
            this.canvas.height = this.model.get('_height_canvas')
                // } else {
                // https://developer.mozilla.org/en-US/docs/Web/API/Element/removeAttribute
                // this.canvas.removeAttribute('_height_canvas');
                // }

            // Update CSS display width and height
            if (this.model.get('width') !== undefined) {
                this.canvas.style.width = this.model.get('width') + 'px'
            } else {
                // https://developer.mozilla.org/en-US/docs/Web/API/Element/removeAttribute
                this.canvas.style.removeAttribute('width');
            }

            if (this.model.get('height') !== undefined) {
                this.canvas.style.height = this.model.get('height') + 'px'
            } else {
                // https://developer.mozilla.org/en-US/docs/Web/API/Element/removeAttribute
                this.canvas.style.removeAttribute('height');
            }

            // Image rendering quality
            // https://developer.mozilla.org/en/docs/Web/CSS/image-rendering
            // auto, crisp-edges, pixelated
            if (this.model.get('pixelated')) {
                this.canvas.style.imageRendering = 'pixelated'
            } else {
                this.canvas.style.imageRendering = 'auto'
            }

            // Draw it!
            this.draw()

            return CanvasImageView.__super__.update.apply(this);
        },

        update_encoded: function() {
            // Python --> JavaScript
            var value = this.model.get('_encoded_url');
            if (value != '') {

                // Load encoded image data into worker image object.
                // this.imageWork.src = 'data:image/' + this.model.get('_format') + ';base64,' + value
                this.imageWork.src = value

                // Event processing and image decoding continues inside imageWork's onload() event
                // handler, which in turn calls this.draw() defined further below.
            }
        },

        something: function() {
            // Testing ideas and debugging problems.

            console.log('something')

            // http://stackoverflow.com/questions/5694986/how-can-i-clone-an-image-in-javascript
            // http://stackoverflow.com/questions/8467121/can-i-copy-a-javascript-image-object
            // http://www.w3schools.com/jsref/met_node_clonenode.asp

            // Multiple options for getting image data
            // var imageFinal = this.imageWork.cloneNode()
            // var data_url = this.canvas.toDataURL()
            // var width = this.model.get('width')
            // var height = this.model.get('height')

            // var image_static = new Image();

            // Find pre-rendered static image place holder.
            var n = 'static_image_' + this.model.get('_uuid')
            var cell = findAncestor(this.canvas, 'cell')
            var image_static = cell.getElementsByClassName(n)[0]; // there should exist just a single output element
            image_static.src = this.canvas.toDataURL();
            image_static.style.display = 'block'

            // Find all rendered instances
            // var elems = document.getElementsByClassName(this.model.get('_uuid'));
            // console.log(elems)
            // for (var i = 0; i < elems.length; i++) {
            //     var elem_canvas = elems[i]
            //     console.log(elem_canvas)
            // }

            // // Build new element output_area.  Alll these details learned by browsing rendered
            // // notebook and copying observed styles.  Quite time consuming!  Blah!
            // var output_area = document.createElement('div')
            // output_area.className = 'output_area'

            // var prompt = document.createElement('div')
            // prompt.className = 'prompt'

            // var output_subarea = document.createElement('div')
            // output_subarea.className = 'output_subarea output_jpeg output_execute_result'
            // // output_png output_result
            // // output_html rendered_html


            // // Assemble output_area
            // output_area.appendChild(prompt)
            // output_area.appendChild(output_subarea)
            // output_subarea.appendChild(image_static)

            // // Now its time to place the above finished output_area in the right spot.

            // // Attach new output_area as first child of current cell's output element.
            // var cell = findAncestor(this.canvas, 'cell')
            // var output = cell.getElementsByClassName('output')[0]; // there should exist just a single output element

            // // New element `output_area` needs to be placed as first child of element `output` in
            // // order to replicate appearance of displayed widgets.  Potential problem is that there
            // // may or may not exist other sibling `output_area` elements.

            // // Pre-existing output_area elements?
            // var existing_output_areas = output.getElementsByClassName('output_area')
            // if (existing_output_areas.length > 0) {
            //     // Place in front of pre-existing output_area elements
            //     output.insertBefore(output_area, existing_output_areas[0]);
            // } else {
            //     // Add as only child
            //     output.appendChild(output_area);
            // }


            // Copy style to static image.
            // https://developer.mozilla.org/en-US/docs/Web/API/Window/getComputedStyle
            // var completeStyle = window.getComputedStyle(this.canvas, null).cssText
            // image_static.style.cssText = completeStyle


            // document.getElementById("placehere").appendChild(image);
            // need to add to current cell's output element:
            // - output area
            //     - output_subarea, output_jpeg (or output_png)



            // See IPython.display.Image for max-width confinement example

            // http://stackoverflow.com/questions/7802744/adding-an-img-element-to-a-div-with-javascript
            // this.appropriate_cell_area['text/html'] = '<img src="' + dataURL + '" width="' + width + '">';

            // IPython.keyboard_manager.enable()
            // $(fig.parent_element).html('<img src="' + dataURL + '" width="' + width + '">');
        },

        remove: function() {
            // Widget is about to be removed from the front end display.

            // This is a place where I could potentially add functionality to replace canvas with a
            // static image. See matplotlib's nbagg for ideas:
            // https://github.com/matplotlib/matplotlib/blob/master/lib/matplotlib/backends/web_backend/nbagg_mpl.js
            // http://stackoverflow.com/questions/7802744/adding-an-img-element-to-a-div-with-javascript
            // http://stackoverflow.com/questions/5694986/how-can-i-clone-an-image-in-javascript
            // http://www.w3schools.com/jsref/met_node_clonenode.asp
            // http://stackoverflow.com/questions/8467121/can-i-copy-a-javascript-image-object


            CanvasImageView.__super__.remove.apply(this, arguments);
        },

        clear: function() {
            // Clear the canvas while preserving current geometry state.
            // http://stackoverflow.com/a/6722031/282840
            var context = this.canvas.getContext('2d');

            context.save();

            context.setTransform(1, 0, 0, 1, 0, 0);
            context.clearRect(0, 0, this.canvas.width, this.canvas.height);

            context.restore();
        },

        draw: function() {
            // Draw image data from internal Image object to the Canvas element.
            // http://www.w3.org/TR/2014/CR-2dcontext-20140821/#drawing-images-to-the-canvas

            // Clear any prior image data.
            this.clear();

            // Update current transform information.
            // -= NOT YET IMPLEMENTED =-

            // Draw image to screen
            var context = this.canvas.getContext('2d');
            context.drawImage(this.imageWork, 0, 0);
        },

        /////////////////////////////////////////
        // JavaScript --> Python
        // Tell Backbone how to respond to JavaScript-generated events
        // Great reference: https://developer.mozilla.org/en-US/docs/Web/Reference/Events
        events: {
            mousemove: 'handle_mouse_move',
            mouseup: 'handle_mouse_generic',
            mousedown: 'handle_mouse_generic',
            wheel: 'handle_mouse_generic',
            click: 'handle_mouse_generic',
            // mouseenter: 'XXX',  // don't worry about these events for now
            // mouseleave: 'XXX',
            // mouseout:   'XXX',
            // mouseover:  'XXX',
        },

        build_mouse_event: function(jev) {
            // Build event data structure to be sent to Python backend

            // Mouse button events
            // https://developer.mozilla.org/en-US/docs/Web/API/MouseEvent/buttons
            //
            // Canvas-local XY coordinates:
            // http://stackoverflow.com/questions/17130395/canvas-html5-real-mouse-position
            // https://developer.mozilla.org/en-US/docs/Web/API/Element.getBoundingClientRect
            var rect = this.canvas.getBoundingClientRect();

            // Convert event types to more practical values for user's Python callback functions.
            var js_to_py = {
                'mousemove': 'move',
                'mouseup': 'up',
                'mousedown': 'down',
                'click': 'click',
                'wheel': 'wheel',
            }

            // Build the event
            var ev = {
                type: js_to_py[jev.originalEvent.type],
                canvasX: parseInt(jev.originalEvent.clientX - rect.left),
                canvasY: parseInt(jev.originalEvent.clientY - rect.top),
                shiftKey: jev.originalEvent.shiftKey,
                altKey: jev.originalEvent.altKey,
                ctrlKey: jev.originalEvent.ctrlKey,
                timeStamp: jev.originalEvent.timeStamp,
                buttons: jev.originalEvent.buttons,
            }

            // Check for `wheel` event
            // https://developer.mozilla.org/en-US/docs/Web/Reference/Events/wheel
            if (jev.originalEvent.type == 'wheel') {
                ev.deltaMode = jev.originalEvent.deltaMode
                ev.deltaX = jev.originalEvent.deltaX
                ev.deltaY = jev.originalEvent.deltaY
                ev.deltaZ = jev.originalEvent.deltaZ
            }

            return ev
        },

        check_mouse_throttle: function(jev) {
            // Return true if enough time has passed
            var delta = jev.originalEvent.timeStamp - this._mouse_timestamp
            if (delta >= this._mouse_time_threshold) {
                this._mouse_timestamp = jev.originalEvent.timeStamp
                return true
            } else {
                return false
            }
        },

        handle_mouse_generic: function(jev) {
            // Generic mouse event handler
            // https://developer.mozilla.org/en-US/docs/Web/Reference/Events
            if (this.model.get('_mouse_active')) {
                var ev = this.build_mouse_event(jev);
                this.model.set('_mouse_event', ev);
                this.touch(); // Must call after any modifications to Backbone Model data.
            }
        },

        handle_mouse_move: function(jev) {
            // Mouse motion event handler
            if (this.model.get('_mouse_active')) {
                if (this.check_mouse_throttle(jev)) {
                    // This event appears to generate a lot of CPU usage.  Throttling is my
                    // simple attempt to mitigate the issue.
                    var ev = this.build_mouse_event(jev);
                    this.model.set('_mouse_event', ev);
                    this.touch(); // Must call after any modifications to Backbone Model data.
                }
            }
        },
    });

    // Simple way to load JS stuff.
    manager.WidgetManager.register_widget_view('CanvasImageView', CanvasImageView);

    // // Official and complicated way to load/register JS stuff.
    // return {
    //     CanvasImageView: CanvasImageView,
    // };
});
